File: java/src/main/java/vsdk/toolkit/common/commandLineOptions/OptionBase.java
package vsdk.toolkit.common.commandLineOptions;

public class OptionBase {
    @FunctionalInterface
    public interface ConsumesValueFunction {
        int apply(Object option);
    }

    @FunctionalInterface
    public interface ApplyFunction {
        boolean apply(Object option, Object context, int argc, String[] argv);
    }

    private String name_;
    private int abbreviationLength_;
    private ConsumesValueFunction consumesValue_;
    private ApplyFunction apply_;
    private Object option_;

    public OptionBase() {
        name_ = null;
        abbreviationLength_ = 0;
        consumesValue_ = null;
        apply_ = null;
        option_ = null;
    }

    public OptionBase(
        String name,
        int abbreviationLength,
        ConsumesValueFunction consumesValue,
        ApplyFunction apply,
        Object option)
    {
        name_ = name;
        abbreviationLength_ = abbreviationLength;
        consumesValue_ = consumesValue;
        apply_ = apply;
        option_ = option;
    }

    public boolean isConfigured() {
        return name_ != null && apply_ != null;
    }

    public String getName() {
        return name_;
    }

    public int getAbbreviationLength() {
        return abbreviationLength_;
    }

    public int consumesValue() {
        if (consumesValue_ == null) {
            return 1;
        }
        return consumesValue_.apply(option_);
    }

    public boolean apply(Object context, int argc, String[] argv) {
        if (apply_ == null) {
            return false;
        }
        return apply_.apply(option_, context, argc, argv);
    }
}

File: typescript/src/main/vsdk/toolkit/common/commandLineOptions/OptionBase.ts
export type ConsumesValueFunction = (option: unknown) => number;
export type ApplyFunction = (
  option: unknown,
  context: unknown,
  argc: number,
  argv: string[] | null
) => boolean;

export class OptionBase {
  private name_: string | null;
  private abbreviationLength_: number;
  private consumesValue_: ConsumesValueFunction | null;
  private apply_: ApplyFunction | null;
  private option_: unknown;

  public constructor();
  public constructor(
    name: string,
    abbreviationLength: number,
    consumesValue: ConsumesValueFunction,
    apply: ApplyFunction,
    option: unknown
  );
  public constructor(
    name?: string,
    abbreviationLength = 0,
    consumesValue: ConsumesValueFunction | null = null,
    apply: ApplyFunction | null = null,
    option: unknown = null
  ) {
    this.name_ = name ?? null;
    this.abbreviationLength_ = abbreviationLength;
    this.consumesValue_ = consumesValue;
    this.apply_ = apply;
    this.option_ = option;
  }

  public isConfigured(): boolean {
    return this.name_ !== null && this.apply_ !== null;
  }

  public getName(): string | null {
    return this.name_;
  }

  public getAbbreviationLength(): number {
    return this.abbreviationLength_;
  }

  public consumesValue(): number {
    if (this.consumesValue_ === null) {
      return 1;
    }
    return this.consumesValue_(this.option_);
  }

  public apply(context: unknown, argc: number, argv: string[] | null): boolean {
    if (this.apply_ === null) {
      return false;
    }
    return this.apply_(this.option_, context, argc, argv);
  }
}
